README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fixedincome <img src="man/figures/logo.png" align="right" width="120" />

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/fixedincome)](https://CRAN.R-project.org/package=fixedincome)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/wilsonfreitas/R-fixedincome/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/wilsonfreitas/R-fixedincome/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

Calculations involving interest rates are usually very easy and
straightforward, but sometimes it involves specific issues that makes
the task of writing structured and reproducible code for it chalenging
and annoying. The `fixedincome` package brings many functions to
strucutre and create facilities to handle with interest rates, term
structure of interest rates and specific issues regarding compounding
rates and day count rules, for example.

Below there are a few examples on how to create and make calculations
with interest rates using `fixedincome`.

## Installation

You can install from CRAN with:

``` r
install.packages("fixedincome")
```

You can install the development version of fixedincome from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("wilsonfreitas/R-fixedincome")
```

## Examples

To create an interest rate we need to specify 4 elements:

-   the value of the interest rate itself, a decimal number
-   the compounding regime of interest rate, that can be `simple`,
    `discrete` or `continuous`.
-   the day count rule which defines how interest is accrued over time,
    we have a few options, for example, `actual/360` where the days
    between two dates are calculated as the difference and the year is
    assumed to be 360 days.
-   the calendar used to count the number of days between two dates, we
    have `actual` calendar that compute the difference between two
    dates.

There is another important topic that wasn’t declared here that is the
*frequency* of interest. To start with the things simple `fixedincome`
handles only with *annual* rates since this represents the great
majority of rates used in financial market contracts, but this
restriction can be reviewed in the future.

Given that let’s declare an annual spot rate with a `simple`
compounding, an `actual/360` and the `actual` calendar.

``` r
library(fixedincome)
sr <- spotrate(0.06, "simple", "actual/360", "actual")
sr
#> [1] "0.06 simple actual/360 actual"
```

Compound the spot rate for 7 months.

``` r
compound(sr, 7, "months")
#> [1] 1.035
```

Also compound using dates.

``` r
compound(sr, as.Date("2022-02-23"), as.Date("2022-12-28"))
#> [1] 1.051333
```

Spot rates can be put inside data.frames.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:fixedincome':
#> 
#>     first, last
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(fixedincome)

df <- tibble(
  rate = spotrate(rep(10.56 / 100, 5),
    compounding = "discrete",
    daycount = "business/252",
    calendar = "Brazil/ANBIMA"
  ),
  terms = term(1:5, "years")
)

df
#> # A tibble: 5 x 2
#>   rate                                       terms  
#>   <SpotRate>                                 <Term> 
#> 1 0.1056 discrete business/252 Brazil/ANBIMA 1 year 
#> 2 0.1056 discrete business/252 Brazil/ANBIMA 2 years
#> 3 0.1056 discrete business/252 Brazil/ANBIMA 3 years
#> 4 0.1056 discrete business/252 Brazil/ANBIMA 4 years
#> 5 0.1056 discrete business/252 Brazil/ANBIMA 5 years
```

The tidyverse verbs can be easily used with `SpotRate` and `Term`
classes.

``` r
df |> mutate(fact = compound(rate, terms))
#> # A tibble: 5 x 3
#>   rate                                       terms    fact
#>   <SpotRate>                                 <Term>  <dbl>
#> 1 0.1056 discrete business/252 Brazil/ANBIMA 1 year   1.11
#> 2 0.1056 discrete business/252 Brazil/ANBIMA 2 years  1.22
#> 3 0.1056 discrete business/252 Brazil/ANBIMA 3 years  1.35
#> 4 0.1056 discrete business/252 Brazil/ANBIMA 4 years  1.49
#> 5 0.1056 discrete business/252 Brazil/ANBIMA 5 years  1.65
```

### Spot rate curves

Let’s create a spot rate curve using web scraping (from B3 website)

``` r
source("examples/utils-functions.R")
#> 
#> Attaching package: 'bizdays'
#> The following object is masked from 'package:stats':
#> 
#>     offset
curve <- get_curve_from_web("2022-02-23")
curve
#>          SpotRateCurve
#> 1 day           0.1065
#> 3 days          0.1064
#> 25 days         0.1111
#> 44 days         0.1138
#> 66 days         0.1168
#> 87 days         0.1189
#> 108 days        0.1207
#> 131 days        0.1220
#> 152 days        0.1227
#> 172 days        0.1235
#> # ... with 29 more rows
#> discrete business/252 Brazil/ANBIMA 
#> Reference date: 2022-02-23
```

`SpotRateCurve` plots can be easily done by calling `plot`.

``` r
plot(curve)
```

<img src="man/figures/README-unnamed-chunk-10-1.png" width="100%" />

For another date.

``` r
curve <- get_curve_from_web("2011-02-23")
plot(curve)
```

<img src="man/figures/README-unnamed-chunk-11-1.png" width="100%" />

It can show the forward rates for the short term by selecting the first
two years.

``` r
curve <- get_curve_from_web("2022-02-23")
plot(fixedincome::first(curve, "2 years"), show_forward = TRUE)
```

<img src="man/figures/README-unnamed-chunk-12-1.png" width="100%" />

Once interpolation is set, it can be used in the plot.

``` r
curve_2y <- fixedincome::first(curve, "2 years")
interpolation(curve_2y) <- interp_flatforward()
plot(curve_2y, use_interpolation = TRUE, legend_location = "bottomright")
```

<img src="man/figures/README-unnamed-chunk-13-1.png" width="100%" />

Parametric models like the Nelson-Siegel-Svensson model can be fitted to
the curve.

``` r
beta1 <- as.numeric(fixedincome::last(curve, "1 day"))
beta2 <- as.numeric(curve[1]) - beta1
interpolation(curve) <- fit_interpolation(
  interp_nelsonsiegelsvensson(beta1, beta2, 0.01, 0.01, 2, 1), curve
)

interpolation(curve)
#> <Interpolation: nelsonsiegelsvensson > 
#>  Parameters:
#>   beta1   beta2   beta3   beta4 lambda1 lambda2 
#>   0.119  -0.013   1.000  -0.975   1.195   1.122
```

Once set to the curve it is used in the plot to show daily forward
rates.

``` r
plot(curve, use_interpolation = TRUE, show_forward = TRUE, legend_location = "bottom")
```

<img src="man/figures/README-unnamed-chunk-15-1.png" width="100%" />

The interpolation can be changed in order to compare different
interpolations and the effects in forward rates.

``` r
interpolation(curve) <- interp_flatforward()
plot(
  curve,
  use_interpolation = TRUE, show_forward = TRUE,
  legend_location = "bottomright"
)
```

<img src="man/figures/README-unnamed-chunk-16-1.png" width="100%" />

Interpolation enables the creation of standardized curves, commonly used
in risk management to build risk factors.

``` r
risk_terms <- c(1, c(3, 6, 9) * 21, c(1, 5, 10) * 252)
risk_curve <- curve[[risk_terms]]
interpolation(risk_curve) <- interp_flatforward()
plot(risk_curve, use_interpolation = TRUE)
```

<img src="man/figures/README-unnamed-chunk-17-1.png" width="100%" />


NEWS.md
# fixedincome 0.0.4

* Improved nelsonsiegel and nelsonsiegelsvensson optimization.
* Corrected bug due to recent changes in `match` in R-devel

# fixedincome 0.0.3

* ggplot2 support.
* Added optional refdate attribute to ForwardRate class. (Issue #29)
* New vignette "Plotting Spot Rate Curves"
* Updated vignette "Spot Rate Curve Interpolation"

# fixedincome 0.0.2

## Term class

* Term class accepts multiple units.
* `toyears` returns Term objects
* New methods `tomonths` and `todays` for Term conversion in different units.

## SpotRate

* SpotRate Compare method split into spectific methods: ==, !=, >, <, >=, <=. The methods >, <, >=, <= raise error for objects with different slots: compounding, daycount, calendar.
* SpotRate Arith methods raise error for objects with different slots: compounding, daycount, calendar.

## SpotRateCurve

* The curve methods `[` and `[[` don't propagate interpolation. The returned curves don't have interpolation. This avoids recurrent `prepare_interpolation` calls.
* New SpotRateCurve method `[[` for missing returns interpolated curve in curve's terms.
* SpotRateCurve converts terms to `days` units, so `days` units becomes the default units to SpotRateCurve.

## Interpolation

* Added `progagate` slot to `Interpolation` class. It defines if the interpolation method is propagated to interpolated curves.
* The `interpolate` function, for NelsonSiegel and NelsonSiegelSvensson models, converts `term` argument to years with `toyears`.
* Organized nelsonsiegel and nelsonsiegelsvensson code.
  * New objective functions (and their gradients) have been created to improve the optimization.

## Compounding

* `rates` method has been renamed to `implied_rate`

## Misc

* Code refactoring
* Updated to bizdays version 0.1.10 (Issue #21)

# fixedincome 0.0.1

* First CRAN version.

R/spotrate-class.R
#' SpotRate class
#'
#' @description
#' The \code{SpotRate} class abstracts the interst rate and has methods
#' to handle many calculations on it.
#'
#' @note
#' The \code{SpotRate} objects are annual rates.
#'
#' @details
#' The \code{SpotRate} class fully specifies spot rates.
#' It has:
#' \itemize{
#'   \item the spot rate values which are numeric values representing the rate.
#'   \item the compounding regime that specifies how to compound the spot
#'         rate. This is a \code{Compounding} object.
#'   \item the daycount rule to compute the compounding periods right
#'         adjusted to the spot rate frequency.
#'   \item the calendar according to which the number of days are counted.
#' }
#'
#' The \code{SpotRate} class is a `numeric`, that represents the
#' interest rate and that has the slots: \code{compounding}, \code{daycount}
#' and \code{calendar}.
#'
#' For example, an annual simple interest rate of 6%, that compounds in
#' calendar days, is defined as follows:
#'
#' ```{r}
#' sr_simple <- spotrate(0.06, "simple", "actual/360", "actual")
#' sr_simple
#' ```
#'
#' `actual/360` is the daycount rule and `actual` is the calendar.
#'
#' Differently, an annual compound interest rate of 10%, that compounds
#' in business days according to calendar `Brazil/ANBIMA` is
#'
#' ```{r}
#' sr_disc <- spotrate(0.1, "discrete", "business/252", "Brazil/ANBIMA")
#' sr_disc
#' ```
#'
#' The \code{calendar} slot is a \code{bizdays} calendar.
#'
#' An $100,000 investment in an instrument that pays that interst rate for 5
#' years has the future value.
#'
#' ```{r}
#' 100000 * compound(sr_disc, term(5, "years"))
#' ```
#'
#' for the simple interest rate we have
#'
#' ```{r}
#' 100000 * compound(sr_simple, term(5, "years"))
#' ```
#'
#' SpotRate objects can be created with vectors
#'
#' ```{r}
#' rates <- c(1.69, 0.16, 0.07, 0.72, 0.10, 1.60, 0.18, 1.56, 0.60, 1.69)
#' sr_vec <- spotrate(rates, "discrete", "business/252", "Brazil/ANBIMA")
#' sr_vec
#' ```
#'
#' and can be put into a `data.frame`
#'
#' ```{r}
#' data.frame(spot_rate = sr_vec)
#' ```
#'
#' once in a `data.frame`, dplyr verbs can be used to manipulate it.
#'
#' ```{r}
#' require(dplyr, warn.conflicts = FALSE)
#'
#' data.frame(spot_rate = sr_vec) |>
#'    mutate(comp = compound(spot_rate, term(5, "months")))
#' ```
#'
#' SpotRate is `numeric`, so it executes arithmetic and comparison operations
#' with `numeric` objects.
#'
#' ```{r}
#' data.frame(spot_rate = sr_vec) |>
#'    mutate(
#'      new_spot_rate = spot_rate + 0.02,
#'      check_gt_1pp = spot_rate > 0.01,
#'      check_gt_nsr = spot_rate > new_spot_rate
#'    )
#' ```
#'
#' SpotRate vectors also are created with the concatenation function `c`.
#'
#' ```{r}
#' c(sr_disc, 0.1, 0.13, 0.14, 0.15)
#' ```
#'
#' Furtherly, all indexing operations of numeric objects are supported by
#' SpotRate objects.
#'
#' ### Invalid Operations
#'
#' Operations involving SpotRate objects with different `compounding`,
#' `daycount` or `calendar`, raise errors.
#'
#' This happens with the following operations:
#'
#' - Compare: >, <, <=, >=
#' - Arithmetic: +, -, *, /
#' - Concatenation: `c`
#'
#' ```{r}
#' try(sr_simple + sr_disc)
#' try(sr_simple > sr_disc)
#' try(c(sr_simple, sr_disc))
#' ```
#'
#' @export
setClass(
  "SpotRate",
  slots = c(
    compounding = "Compounding",
    daycount = "Daycount",
    calendar = "character"
  ),
  contains = "numeric"
)

#' Create SpotRate objects
#'
#' @description
#' `spotrate()` function creates `SpotRate` objects.
#'
#' @param x a numeric vector representing spot rate values.
#' @param compounding a \code{Compounding} object.
#' @param daycount a \code{Daycount} object.
#' @param calendar a \code{bizdays} calendar.
#' @param .copyfrom a \code{SpotRate} object used as reference to copy
#'        attributes.
#'
#' @return A `SpotRate` object.
#'
#' @examples
#' spotrate(0.06, "continuous", "actual/365", "actual")
#' spotrate(c(0.06, 0.07, 0.08), "continuous", "actual/365", "actual")
#' @export
spotrate <- function(x, compounding, daycount, calendar, .copyfrom = NULL) {
  if (!is.null(.copyfrom)) {
    x <- if (missing(x)) .copyfrom@.Data else x
    compounding <- if (missing(compounding)) {
      .copyfrom@compounding
    } else {
      compounding
    }
    daycount <- if (missing(daycount)) .copyfrom@daycount else daycount
    calendar <- if (missing(calendar)) .copyfrom@calendar else calendar
  }

  compounding <- if (is.character(compounding)) {
    compounding(compounding)
  } else {
    compounding
  }
  daycount <- if (is.character(daycount)) daycount(daycount) else daycount
  new("SpotRate", x,
    compounding = compounding, daycount = daycount,
    calendar = calendar
  )
}

#' @export
as.character.SpotRate <- function(x, ...) {
  paste(
    x@.Data, as(x@compounding, "character"),
    as(x@daycount, "character"), x@calendar
  )
}

#' @export
format.SpotRate <- function(x, ...) {
  hdr <- paste(
    as(x@compounding, "character"),
    as(x@daycount, "character"), x@calendar
  )
  paste(callGeneric(x@.Data, ...), hdr)
}

#' @export
setMethod(
  "show",
  signature(object = "SpotRate"),
  function(object) {
    print(format(object))
  }
)

#' @export
`[.SpotRate` <- function(x, i) {
  spotrate(x@.Data[i], x@compounding, x@daycount, x@calendar)
}

#' @export
`[<-.SpotRate` <- function(x, i, value) {
  x@.Data[i] <- value
  x
}

#' @export
c.SpotRate <- function(x, ...) {
  dots <- list(...)
  nempty <- sapply(dots, length) != 0
  elements <- lapply(dots[nempty], spr_builder(x))
  values_ <- c(x@.Data, unlist(lapply(elements, as.numeric)))
  spotrate(values_, x@compounding, x@daycount, x@calendar)
}

#' Coerce to SpotRate
#'
#' @description
#' Coerce character objects to SpotRate class
#'
#' @param x a character with SpotRate specification.
#' @param simplify a boolean indicating whether to simplify SpotRate creation
#'        or not. Defaults to TRUE.
#' @param ... additional arguments
#'
#' @details
#'
#' The character representation of a SpotRate is as follows:
#'
#' \preformatted{"RATE COMPOUNDING DAYCOUNT CALENDAR"}
#'
#' where:
#' \itemize{
#'   \item \code{RATE} is a numeric value
#'   \item \code{COMPOUNDING} is one of the following:
#'         \code{simple}, \code{discrete}, \code{continuous}
#'   \item \code{DAYCOUNT} is a valid day count rule, pex. \code{business/252},
#'         see [Daycount-class].
#'   \item \code{CALENDAR} is the name of a bizdays calendar.
#' }
#'
#' \code{simplify} check if compounding, daycount and calendar are the same for
#' all given characters.
#' If it is true the returned object is a SpotRate otherwise a \code{list} with
#' SpotRate objects is returned.
#'
#' @return A `SpotRate` object created from a string.
#'
#' @examples
#'
#' as.spotrate(c(
#'   "0.06 simple actual/365 actual",
#'   "0.11 discrete business/252 actual"
#' ))
#' @export
setGeneric(
  "as.spotrate",
  function(x, ...) {
    standardGeneric("as.spotrate")
  }
)

.parse_spotrate <- function(x) {
  lapply(strsplit(x, "\\s+", perl = TRUE), function(x) {
    if (length(x) != 4) {
      stop("Invalid spotrate specification")
    }
    spotrate(as.numeric(x[1]), x[2], x[3], x[4])
  })
}

#' @rdname as.spotrate
#' @export
setMethod(
  "as.spotrate",
  "character",
  function(x, simplify = TRUE) {
    if (simplify) {
      m <- regexec("^(\\d+\\.\\d+)\\s+(.*)$", x)
      rm <- regmatches(x, m)
      specs <- unique(sapply(rm, function(x) x[3]))
      if (length(specs) == 1) {
        value <- sapply(rm, function(x) as.numeric(x[2]))
        specs <- strsplit(specs, "\\s+")[[1]]
        if (length(specs) != 3) {
          stop("Invalid spotrate specification")
        }
        spotrate(value, specs[1], specs[2], specs[3])
      } else {
        .parse_spotrate(x)
      }
    } else {
      .parse_spotrate(x)
    }
  }
)

#' @export
setMethod(
  "Arith",
  signature(e1 = "SpotRate", e2 = "SpotRate"),
  function(e1, e2) {
    e1@.Data <- callGeneric(e1@.Data, e2@.Data)
    stop_if_spotrate_slots_differ(
      e1, e2,
      "SpotRate objects have different slots"
    )
    e1
  }
)

#' @export
setMethod(
  "Arith",
  signature(e1 = "SpotRate", e2 = "numeric"),
  function(e1, e2) {
    e1@.Data <- callGeneric(e1@.Data, e2)
    e1
  }
)

#' @export
setMethod(
  "Arith",
  signature(e1 = "numeric", e2 = "SpotRate"),
  function(e1, e2) {
    e2@.Data <- callGeneric(e1, e2@.Data)
    e2
  }
)

#' SpotRate comparison operations
#'
#' Comparison operations with SpotRate class
#' \code{SpotRate} objects can be compared among themselves or with numeric
#' variables.
#'
#' @param e1 a \code{SpotRate} object or a numeric
#' @param e2 a \code{SpotRate} object or a numeric
#'
#' @return
#' A boolean `logical` object.
#' The comparison with `SpotRate` objects only takes all fields
#' into account.
#' Comparing `SpotRate` against numeric values is equivalent to
#' coerce the `SpotRate` object to numeric execute the operation,
#' this is a syntax sugar for a shortcut that is commonly applied.
#' @name spotrate-compare-method
#'
#' @examples
#'
#' spr <- as.spotrate("0.06 simple actual/365 actual")
#' spr == 0.06
#' spr != 0.05
#' spr > 0.05
#' spr < 0.1
#' spr >= 0.05
#' spr <= 0.1
#'
#' spr1 <- spotrate(0.06, "simple", "actual/365", "actual")
#' spr2 <- spotrate(0.02, "simple", "actual/365", "actual")
#' spr1 == spr2
#' spr1 != spr2
#' spr1 > spr2
#' spr1 < spr2
#' spr1 >= spr2
#' spr1 <= spr2
#'
#' # compare spotrate with different slots
#' spr2 <- spotrate(0.06, "discrete", "actual/365", "actual")
#' spr1 == spr2
#' spr1 != spr2
#' try(spr1 > spr2)
#' try(spr1 < spr2)
#' try(spr1 >= spr2)
#' try(spr1 <= spr2)
#'
NULL

#' @rdname spotrate-compare-method
#' @export
setMethod(
  ">=",
  signature(e1 = "SpotRate", e2 = "SpotRate"),
  function(e1, e2) {
    stop_if_spotrate_slots_differ(e1, e2, "SpotRate objects have different slots")
    callGeneric(e1@.Data, e2@.Data)
  }
)

#' @rdname spotrate-compare-method
#' @export
setMethod(
  "<=",
  signature(e1 = "SpotRate", e2 = "SpotRate"),
  function(e1, e2) {
    stop_if_spotrate_slots_differ(e1, e2, "SpotRate objects have different slots")
    callGeneric(e1@.Data, e2@.Data)
  }
)

#' @rdname spotrate-compare-method
#' @export
setMethod(
  "<",
  signature(e1 = "SpotRate", e2 = "SpotRate"),
  function(e1, e2) {
    stop_if_spotrate_slots_differ(e1, e2, "SpotRate objects have different slots")
    callGeneric(e1@.Data, e2@.Data)
  }
)

#' @rdname spotrate-compare-method
#' @export
setMethod(
  ">",
  signature(e1 = "SpotRate", e2 = "SpotRate"),
  function(e1, e2) {
    stop_if_spotrate_slots_differ(e1, e2, "SpotRate objects have different slots")
    callGeneric(e1@.Data, e2@.Data)
  }
)

#' @rdname spotrate-compare-method
#' @export
setMethod(
  "==",
  signature(e1 = "SpotRate", e2 = "SpotRate"),
  function(e1, e2) {
    callGeneric(e1@.Data, e2@.Data) & check_slots(e1, e2)
  }
)

#' @rdname spotrate-compare-method
#' @export
setMethod(
  "!=",
  signature(e1 = "SpotRate", e2 = "SpotRate"),
  function(e1, e2) {
    callGeneric(e1@.Data, e2@.Data) | !check_slots(e1, e2)
  }
)

#' @rdname spotrate-compare-method
#' @export
setMethod(
  "Compare",
  signature(e1 = "SpotRate", e2 = "numeric"),
  function(e1, e2) {
    callGeneric(e1@.Data, e2)
  }
)

#' @rdname spotrate-compare-method
#' @export
setMethod(
  "Compare",
  signature(e1 = "numeric", e2 = "SpotRate"),
  function(e1, e2) {
    callGeneric(e1, e2@.Data)
  }
)

# support functions ----

check_slots <- function(e1, e2) {
  (e1@compounding == e2@compounding) &
    (e1@daycount == e2@daycount) &
    (e1@calendar == e2@calendar)
}

warn_if_spotrate_slots_differ <- function(e1, e2, msg) {
  if (!check_slots(e1, e2)) {
    warning(msg)
  }
}

stop_if_spotrate_slots_differ <- function(e1, e2, msg) {
  if (!check_slots(e1, e2)) {
    stop(msg)
  }
}

spr_builder <- function(x) {
  function(values_) {
    if (is(values_, "SpotRate")) {
      stop_if_spotrate_slots_differ(
        x,
        values_,
        "SpotRate objects have different slots"
      )
      values_ <- as.numeric(values_)
    }
    spotrate(values_, x@compounding, x@daycount, x@calendar)
  }
}
